### PR TITLE
add soft-rasterizer-pytorch example

### PR DESCRIPTION
**Pull request by shihhsinliNV**
This is a port of the slang-torch soft-rasterizer example to SlangPy with PyTorch interop.

- Add ITensor-based kernel soft_rasterizer2d_itensor.slang
- Add main.py training loop using PyTorch (Adam), saves animation

### CODE REVIEW

**Review by jkwak-work — COMMENTED** [review]
I like this a lot.
It seems much simpler than other examples:
- https://github.com/shader-slang/slang-torch/blob/main/examples/soft-rasterizer-example/rasterizer2d.py
- https://github.com/shader-slang/slangpy-samples/blob/main/examples/soft-rasterizer/main.py

I think we also need the document like:
- https://github.com/shader-slang/slang-torch/blob/main/examples/soft-rasterizer-example/README.md
Because after we deprecate slang-torch repo, we want to keep the doc along with the example.

File: examples/soft-rasterizer-pytorch/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Soft Rasterizer (SlangPy + PyTorch)
+
+This is a port of the slang-torch soft-rasterizer example to SlangPy with PyTorch interop.
+- Original example: https://github.com/shader-slang/slang-torch/tree/main/examples/soft-rasterizer-example
+- Because slang-torch is deprecated, this port exists to provide a maintained SlangPy-based version with PyTorch interop.
```
[the remark this answers is not in the document]
Comment (ccummingsNV) [reply examples/soft-rasterizer-pytorch/README.md:5]: I agree - nothing wrong with highlighting where it came from, but discussion of deprecation should be reserved for well thought out/explained info on the web site, not comments in git hub :)

File: examples/soft-rasterizer-pytorch/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+- Note: slangpy-samples already includes a soft-rasterizer example without PyTorch interop (see `examples/soft-rasterizer/`). This sample shows the PyTorch-integrated variant.
+
+## Contents
+- `soft_rasterizer2d_itensor.slang` — per-pixel soft rasterizer using `ITensor<>` for differentiable inputs (vertices, color) and a no-diff camera.
+- `main.py` — host script that renders a target image, optimizes triangle vertices/color with Adam, and saves an animation.
+
+## Requirements
+- CUDA-capable GPU
+- Python packages: `torch`, `slangpy`, `matplotlib`
```
Comment (ccummingsNV) [inline examples/soft-rasterizer-pytorch/README.md:14]: Please add a 'requirements.txt' in the same folder, and explain to users how to run requirements.txt. If the torch package needs to come from a specific index (for cuda) this should also be covered. We want to avoid the user having to guess at the exact setup in order to run an example.

File: examples/soft-rasterizer-pytorch/main.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import torch
+import timeit
+import matplotlib.pyplot as plt
+
+import slangpy as spy
+
+
+def main():
+    W, H = 1024, 1024
+    device = spy.create_torch_device(type=spy.DeviceType.cuda)
```
Comment (ccummingsNV) [inline examples/soft-rasterizer-pytorch/main.py:12]: You should be able to remove the 'type=' here and it stick with CUDA - though check :) 

Probably worth a big comment here that this utility function is the easiest way to make a torch compatible device, and that the default device with best performance is CUDA. Maybe even link to the docs - it's a key detail.

**Review by ccummingsNV — CHANGES_REQUESTED** [review]
A few suggested changes but looks pretty good. There are 2 ways you can validate memory copying:
1. Pretty reliable: debug external\slangpy\slangpy\torchintegration\torchtensormarshall.py - that's where the copies happen
2. Ultra reliable: boot up NSight Systems, capture CUDA events and check how many cuMemCpys occur per dispatch (I think I've got it down to 0)

Maybe get a torch master like @saipraveenb25 just to validate your torch code - make sure you've got the right number of grad requiring and detaching etc. We don't want to confuse people into thinking 'special' torch code is needed for slangpy. Or conversely, if 'special' torch code is required, we need to fix that :)

File: examples/soft-rasterizer-pytorch/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shihhsinliNV) [reply examples/soft-rasterizer-pytorch/README.md:5]: Ack, will remove this statement.

File: examples/soft-rasterizer-pytorch/main.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    # Preallocate output once and reuse each iteration
+    img = torch.empty((W, H, 3), dtype=torch.float32, device='cuda', requires_grad=True)
+
+    def step(i):
+        print(f"Iteration {i}")
+        optim.zero_grad()
+
+        nonlocal img
+        img = img.detach()
+        img.requires_grad_(True)
```
[the remark this answers is not in the document]
Comment (shihhsinliNV) [reply examples/soft-rasterizer-pytorch/main.py:47]: requires_grad_(True) can be removed.
In my testing, detach() is required otherwise img.grad doesn't get updated in each iteration. It seems that detach() allows reuse of the same tensor storage in each iteration as a fresh leaf. It breaks the previous autograd graph so the next iteration doesn’t try to backprop through prior steps or retain old history.

**Comment by shihhsinliNV** [comment]
> I think we also need the document like:
> 
> * https://github.com/shader-slang/slang-torch/blob/main/examples/soft-rasterizer-example/README.md
>   Because after we deprecate slang-torch repo, we want to keep the doc along with the example.
Added more content in README. I'm keeping it minimal since soft-rasterizer already outlined the algorithm.

File: examples/soft-rasterizer-pytorch/README.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
+# Soft Rasterizer (SlangPy + PyTorch)
+
+This is a port of the slang-torch soft-rasterizer example to SlangPy with PyTorch interop.
+- Original example: https://github.com/shader-slang/slang-torch/tree/main/examples/soft-rasterizer-example
+- Note: slangpy-samples already includes a soft-rasterizer example without PyTorch interop (see `examples/soft-rasterizer/`). This sample shows the PyTorch-integrated variant.
+
+## Contents
+- `soft_rasterizer2d_itensor.slang` — per-pixel soft rasterizer using `ITensor<>` for differentiable inputs (vertices, color).
```
[the remark this answers is not in the document]
Comment (shihhsinliNV) [reply examples/soft-rasterizer-pytorch/README.md:8]: I don't think there's any documentation on ITensor. The way I figured out the use of it was reading slangpy tests like https://github.com/shader-slang/slangpy/blob/main/slangpy/tests/slangpy_tests/test_tensor.slang#L48

This actually bring up an important question about whether ITensor(and IRWTensor etc.) is intended as a public interface that developer can use. Maybe @ccummingsNV has more insight into this. 

On a high level, this is my understanding of how ITensor binding works:
1. Python reflection maps ITensor parameters to an ITensor-aware marshaller.
2. When a torch.Tensor is passed, it’s wrapped as a C++ TensorRef; the Python TensorRefMarshall produces the per-parameter call-data (Cuda device pointer, shape, strides, gradients). With the device pointer, the interop can be zero-copy for Cuda backend.

(Note that these are purely from inspecting the code base, I didn't manually perform any code tracing so it might not be entirely accurate).

[the remark this answers is not in the document]
Comment (jkwak-work) [reply examples/soft-rasterizer-pytorch/README.md:8]: Thanks for the information.
I think we can inspect the details later.
But what you described makes sense to me and I believe that must be what is happening.
